compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/generators/ContributedInterfaceSupertypeGenerator.kt

```diff
@@ -311,6 +294,24 @@ internal class ContributedInterfaceSupertypeGenerator(session: FirSession) :
     return pendingRankReplacements
   }
 
+  private fun FirClassLikeSymbol<*>.implicitBoundType(
+    typeResolver: TypeResolveService
+  ): ConeKotlinType {
+    return if (fir.resolveState.resolvePhase == FirResolvePhase.RAW_FIR) {
+        // When processing bindings in the same module or compilation, we need to handle supertypes
+        // that have not been resolved yet
+        (this as FirClassSymbol<*>).fir.superTypeRefs.map {
+          typeResolver.resolveUserType(it as FirUserTypeRef).coneType
+        }
+      } else {
+        (this as FirClassSymbol<*>).resolvedSuperTypes
+      }
+      .singleOrError {
+        val superTypeFqNames = map { it.classId?.asSingleFqName() }.joinToString()
+        "${classId.asSingleFqName()} has a ranked binding with no explicit bound type and $size supertypes ($superTypeFqNames). There must be exactly one supertype or an explicit bound type."
+      }
```

Comment:
Good with merging as-is but these `.fir` refs are probably internal APIs we should be avoiding iirc. Usually there's extension helpers to avoid it. Let me poke around a bit

Reply:
tried a bit, I see the difficulty now, giving up until a later time

Reply:
Aye this section gets a bit tricky. Not sure of alternatives for `fir.superTypeRefs`, but we could probably replace the direct `fir.resolveState` reference by switching to something like a try-catch that first tries to access `resolvedSuperTypes` and falls back to the unresolved path